teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1130 @@
+
+<table>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step1
+spec:
+  image: alpine
+  script: |
+    echo "I am expendable"
+```
+</td>
+<td rowspan=2>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      ref:
+        name: step1
+    - name: s2
+      ref:
+        name: step2
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step2
+spec:
+  image: alpine
+  script: |
+    echo "I am special"
+```
+</td>
+</tr>
+</table>
+
+A `Task` can refer to the `StepAction` from the cluster or fetch it from remote sources via a resolver. 
+
+<table>
+<tr>
+<td>StepAction from cluster</td>
+<td>StepAction from Git repository</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      ref:
+        name: step1
+```
+</td>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      ref:
+        resolver: git
+        params:
+          - name: url
+            value: https://...
+```
+</td>
+</tr>
+</table>
+
+<table>
+<tr>
+<td>Example StepAction</td>
+<td>Example Task with orchestration in Steps</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step1
+spec:
+  image: alpine
+  script: |
+    echo "I am expendable"
+
+```
+</td>
+<td rowspan=2>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      ref:
+        name: step1
+      onError: continue
+    - name: s2
+      ref:
+        name: step2
+      computeResources:
+        requests:
+          memory: 1Gi
+          cpu: 500m
+        limits:
+          memory: 2Gi
+          cpu: 800m
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step2
+spec:
+  image: alpine
+  script: |
+    echo "I am special"
+  securityContext:
+    privileged: true
+```
+</td>
+</tr>
+</table>
+
+#### Remote Resolution
+When it comes to remote resolution, here is the general flow. 
+
+The pipelines controller creates a Resolution request CRD. The resolvers fetch the resource from the remote locations (`git/artifacthub/oci bundle`) and update the CRD with the raw content (i.e. unmarshalled bytes). The `resolvers` are not aware of the exact Struct that the content needs to resolve to.
+The updated to the CRD triggers another reconcile loop and the pipelines controller then accesses the raw content and un-marshalls it to `Pipeline/Task/StepAction`.
+
+This was demonstrated for the `git` resolver in the `StepAction` [POC](https://github.com/tektoncd/pipeline/pull/7124). Assuming that the other resolvers are similar, we do not need to update the resolvers to resolve `StepActions`.
```

Comment:
@vdemeester PTAL at this. It tries to address your feedback from [here](https://docs.google.com/document/d/1z2ME1o_XHvqv6cVEeElljvqVqHV8-XwtXdTkNklFU_8/edit?disco=AAAA7J6Zrh8).

Reply:
Yes. There is no change to do on the resolver's code from pipeline, etc.. But a resolver could definitely only resolve `Pipeline` or `Task` (and refuse resolving any other type requested). In any case, if need be, we'll modify those, it's not a problem 😉